### PR TITLE
Instagram Channel

### PR DESCRIPTION
**Pull request by matmsa27**
Hi guys, I'm opening the PRs to support the Instagram messenger channel on Rapidpro according to #1516 .

PR on [courier](https://github.com/nyaruka/courier/pull/392).
PR on [gocommon](https://github.com/nyaruka/gocommon/pull/55).

### CODE REVIEW

File: temba/channels/types/instagram/tests.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+            None,
+            "IG",
+            name="Instagram",
+            address="019283",
+            role="SR",
+            schemes=["instagram"],
+            config={"auth_token": "09876543", "page_id": "123456"},
+        )
+
+    @override_settings(INSTAGRAM_APPLICATION_ID="IG_APP_ID", INSTAGRAM_APPLICATION_SECRET="IG_APP_SECRET")
```
Comment (norkans7) [inline temba/channels/types/instagram/tests.py:30]: Update to use the same Facebook app as the FBA type

File: temba/channels/types/instagram/type.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+from ...models import Channel, ChannelType
+from .views import ClaimView, RefreshToken
+
+
+class InstagramType(ChannelType):
+    """
+    A Instagram channel
+    """
+
+    extra_links = [dict(name=_("Reconnect Facebook Page"), link="channels.types.instagram.refresh_token")]
```
Comment (norkans7) [inline temba/channels/types/instagram/type.py:17]: Make sure the name here is correct, are we connecting a Facebook page?

File: temba/channels/types/instagram/views.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+from ...models import Channel
+from ...views import ClaimViewMixin
+
+
+class ClaimView(ClaimViewMixin, SmartFormView):
+    class Form(ClaimViewMixin.Form):
+
+        user_access_token = forms.CharField(min_length=32, required=True, help_text=_("The User Access Token"))
+        fb_user_id = forms.CharField(
+            required=True, help_text=_("The Facebook User ID of the admin that connected the channel")
```
Comment (norkans7) [inline temba/channels/types/instagram/views.py:20]: Facebook?

File: temba/channels/types/instagram/views.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+        )
+        page_name = forms.CharField(required=True, help_text=_("The name of the Facebook page"))
+        page_id = forms.IntegerField(required=True, help_text="The Facebook Page ID")
```
Comment (norkans7) [inline temba/channels/types/instagram/views.py:23]: Facebook page?

File: temba/settings_common.py [hunk 5]
```diff
@@ -1219,13 +1220,16 @@
...
 #
-#    4. Generate a Random Secret to use as Facebook Webhook Secret as described
+#    4. Generate a Random Secret to use as Facebook Webhook Secret or Instagram Webhook Secret as described
 #       on https://developers.facebook.com/docs/messenger-platform/webhook#setup
 #
 # -----------------------------------------------------------------------------------
 FACEBOOK_APPLICATION_ID = os.environ.get("FACEBOOK_APPLICATION_ID", "")
 FACEBOOK_APPLICATION_SECRET = os.environ.get("FACEBOOK_APPLICATION_SECRET", "")
 FACEBOOK_WEBHOOK_SECRET = os.environ.get("FACEBOOK_WEBHOOK_SECRET", "")
+INSTAGRAM_APPLICATION_ID = os.environ.get("INSTAGRAM_APPLICATION_ID", "")
```
Comment (norkans7) [inline temba/settings_common.py:1230]: Adjust these to use the existing Facebook app

File: templates/channels/types/instagram/claim.haml [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+        %li
+          -trans "Approve the permissions, these are required for us to send and receive messages on your behalf."
+
+        %li
+          -trans "Finally, select the page you want to add as a channel."
+
+    #fb-app-connect.flex.mt-4
+      .button-primary.connect-facebook
+        -trans "Add Facebook Page"
```
Comment (norkans7) [inline templates/channels/types/instagram/claim.haml:34]: `Add Instagram page`

**Review by norkans7 — CHANGES_REQUESTED** [review]
As on the courier changes adjust to use the same Facebook app and audit these changes to make sure the the text shown to the users are correctly Instagram page and not facebook page to not confuse users

**Review by norkans7 — CHANGES_REQUESTED** [review]
Please audit the changes again I still see places mentioning facebook page instead of Instagram business account

**Comment by Robi9** [comment]
@norkans7  In some places I left it as a Facebook page because it is really necessary to log in to Facebook and choose a Facebook page, this page has to be linked to a business Instagram account.

File: templates/channels/types/instagram/refresh_token.haml [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    -if error_connect
+      -trans "Error with token, you need to reconnect the Instagram Business Account by clicking the button below"
+    -else
+      -trans "Everything looks good. No need to reconnect"
+
+-block form
+  .mt-4.card
+    #fb-channel-error.mb-4(style="display:none;")
+      %p
+        -trans "Error reconnecting Facebook page. Please retry"
```
Comment (norkans7) [inline templates/channels/types/instagram/refresh_token.haml:15]: What about this error message here?

**Comment by norkans7** [comment]
Can you share a few screenshots of the pages added?

File: temba/channels/types/instagram/views.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+            required=True, help_text=_("The Facebook User ID of the admin that connected the channel")
+        )
+
+    slug_url_kwarg = "uuid"
+    success_url = "uuid@channels.channel_read"
+    form_class = Form
+    permission = "channels.channel_claim"
+    fields = ()
+    template_name = "channels/types/instagram/refresh_token.html"
+    title = _("Reconnect Facebook Page")
```
Comment (norkans7) [inline temba/channels/types/instagram/views.py:151]: @Robi9 I really think this title should not include Facebook, I think having Instagram there is better to not confuse this view with the other Facebook type

Please share screenshots for the the different screens that way we'll see easily the text that could be confusing

**Comment by Robi9** [comment]
Screenshots:

Screenshot from 2022-01-06 16-13-07 (1)
Screenshot from 2022-01-06 16-13-07
image

**Comment by norkans7** [comment]
Good, those looks well

Can you share a few more when you click on the "Reconnect Instagram Business Account" button?

**Comment by Robi9** [comment]
I clicked on Reconnect Instagram Business Account and there it says "Reconnect Facebook Page", I'll fix it.

**Comment by Robi9** [comment]
@norkans7 After correction:

image

**Comment by norkans7** [comment]
Awesome, looks good now, I'll have another pass and notify if I notice something else

File: temba/contacts/models.py [hunk 9]
```diff
@@ -188,7 +190,7 @@ def validate(cls, urn, country_code=None):
                 return False
 
         # facebook uses integer ids or temp ref ids
-        elif scheme == cls.FACEBOOK_SCHEME:
+        elif scheme in [cls.FACEBOOK_SCHEME]:
```
Comment (norkans7) [inline temba/contacts/models.py:193]: @Robi9 I think we need to add some validation for Instagram URN in this method

I think adding that at https://github.com/rapidpro/rapidpro/blob/44f12b8593231ecb33fc652a5766515fa474bd2f/temba/contacts/models.py#L205 is enough.

Remember to add a test for that too

**Review by norkans7 — CHANGES_REQUESTED** [review]
Almost there, just the URN validation for Instagram

File: temba/contacts/models.py (hunk 9 above)
Comment (norkans7) [reply temba/contacts/models.py]: @Robi9 why not simply add Instagram to the other all digits here and not mix that with facebook one that supports referral as well

**Review by norkans7 — CHANGES_REQUESTED** [review]
remember to apply `./code_checks.py`

**Review by norkans7 — APPROVED** [review]
Looks good Roberta

**Comment by matmsa27** [comment]
Hi @rowanseymour , what do you think? We open to receive your feedbacks to improve this PR.

File: temba/channels/types/instagram/type.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
+import requests
+
+from django.conf.urls import url
```
Comment (rowanseymour) [inline temba/channels/types/instagram/type.py:3]: I know this was just copied from another channel but we've now updated the other channels - should use `from django.urls import re_path` instead. We're cleaning up use of deprecated functions so we can update to Django 4.

Comment (rowanseymour) [inline temba/channels/types/instagram/type.py:3]: I know this was just copied from another channel but we've now updated the other channels - should use `from django.urls import re_path` instead. We're cleaning up use of deprecated functions so we can update to Django 4.

File: temba/channels/types/instagram/type.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+from django.utils.translation import ugettext_lazy as _
```
Comment (rowanseymour) [inline temba/channels/types/instagram/type.py:4]: should be `gettext_lazy`

File: templates/channels/types/instagram/claim.haml [hunk 12]
```diff
@@ -0,0 +1,136 @@
...
+  -blocktrans trimmed with name=brand.name
+    You can connect your Instagram business account to {{name}} in just a few simple steps.
+
+-block form
+  .mt-4.card
+    #fb-guide
+      %ol.steps
+        %li
+          -trans "The page must be linked to your instagram business account, so check how to link"
+          %a{:href => 'https://help.instagram.com/399237934150902'} here
```
Comment (rowanseymour) [inline templates/channels/types/instagram/claim.haml:14]: not familiar with this template syntax.. what is it and does it work?

File: templates/channels/types/instagram/claim.haml [hunk 13]
```diff
@@ -0,0 +1,136 @@
...
+    %form#claim-form(style="display:none;" method="POST" action="{{ claim_url }}")
+      {% csrf_token %}
+      %input#fb-user-id(type="text" name="fb_user_id")
+      %input#user-access-token(type="text" name="user_access_token")
+      %input#page-id(type="text" name="page_id")
+      %input#page-name(type="text" name="page_name")
+
+  .mt-4.card#fb-channel-options(style="display:none;")
+    %p(style="font-size:1rem;")
+      -trans "Select the page you want to add as a channel:"
```
Comment (rowanseymour) [inline templates/channels/types/instagram/claim.haml:48]: is there a screenshot of this part we can see?

**Review by rowanseymour — CHANGES_REQUESTED** [review]

File: templates/channels/types/instagram/claim.haml (hunk 12 above)
Comment (Robi9) [reply templates/channels/types/instagram/claim.haml:14]: It's just including this link in the word "here" and it's working. As shown in the image below, in step 1.

image

File: templates/channels/types/instagram/claim.haml (hunk 13 above)
Comment (Robi9) [reply templates/channels/types/instagram/claim.haml:48]: print

**Review by rowanseymour — APPROVED** [review]

**Pull request merged**